### DISCUSSION TITLE
does arrayjoin support specify tile locations to produce a mosaic image

### DISCUSSION THREAD

**Discussion by zhusihan-python** [libvips/libvips#3423] [Q&A]
hello, i have some microscopy images scanned snakelike
```
↓  ↑→ ↓
↓  ↑  ↓
↓→ ↑  ↓
```
got pictures as below
```
000_000.jpg  001_002.jpg  002_000.jpg
000_001.jpg  001_001.jpg  002_001.jpg
000_002.jpg  001_000.jpg  002_002.jpg
```
and i can get the correct location computed by itk montage library, like below
```
dim = 2
000_000.jpg;;(0, 0)
001_002.jpg;;(w_offset1 , 0)
002_000.jpg;;(w_offset2 , 0)
000_001.jpg;;(0, h_offset1 )
001_001.jpg;;(w_offset3, h_offset2 )
002_001.jpg;;(w_offset4, h_offset3 )
000_002.jpg;;(0, h_offset4)
001_000.jpg;;(w_offset5, h_offset5 )
002_002.jpg;;(w_offset6, h_offset6 )
```
the problem is when itk generate the final mosaic images, it use too much memory, so i want to use libvips to generate the final image with the images location i already got

the vips_arrayjoin only support fixed hspacing and vspacing right, i wander is there a way to specify a location to each tile when i generate a grid mosaic image by livips, similiar to pillow Image .paste(region_specific) on canvas

or i should use join.merge to merge the images one by one like
`VImage(000_000.jpg).merge(000_001.jpg,  'vertical', relative_offset_w, relative_offset_y, mblend)`

**Comment by jcupitt** [comment]
Hello @zhusihan-python,

You can use `insert` to paste images into a big canvas, or `merge` to make a series of pairwise joins with feathered edges.

`composite` will join an array of images and you can give an x/y position for each one, but it'll get slow beyond about 10,000 images (we should fix this). I don't know how many you want to join.

**Reply by jcupitt** [reply]
`insert` will get slow for very large image sets too. How many do you need to join, how large are the tiles, and what format do you have them in?

**Reply by zhusihan-python** [reply]
Thank you for the quick reply.

my dataset is around 2000 4096*3000 rgb jpg images.
the rgb image without a alpha channel, each image will just overwrite the previous one as this [issue](https://github.com/libvips/pyvips/issues/40#issuecomment-386264741) said right

and if i use composite, how does it know my images are snakelike, and how can i give the x/y position for each tile
suppose my code like this
```
import pyvips
import glob

images = glob.glob(r"E:\data\captures_99\flat_*.jpg")

images = [pyvips.Image.new_from_file(filename, access="sequential")
          for filename in images]

image = images[0].composite(images[1:], "over")

image.write_to_file(r"E:\data\captures_99\vips_mosaic.jpg")
```

**Reply by jcupitt** [reply]
2000 images should be fine. I found an old demo prog for you:

```python
!/usr/bin/python3

import sys
import random
import pyvips

# sigma of the feather we add, and approximate radius in pixels
FEATHER_SIGMA = 4
FEATHER_WIDTH = 7 
    
print("loading images ...")
images = [pyvips.Image.new_from_file(filename, access="sequential") for filename in sys.argv[2:]]

def make_white(width, height):
    white = pyvips.Image.black(width, height) + 255
    white = white.cast('uchar')

    return white

# make a one-band uchar image with a feathered edge
def soft_alpha(width, height):
    black = pyvips.Image.black(width, height) 
    white = make_white(width - 2 * FEATHER_WIDTH, 
                       height - 2 * FEATHER_WIDTH)
    mask = black.insert(white, FEATHER_WIDTH, FEATHER_WIDTH)
    alpha = mask.gaussblur(FEATHER_SIGMA, precision='integer')
    white = make_white(width - 4 * FEATHER_WIDTH, 
                       height - 4 * FEATHER_WIDTH)
    alpha = alpha.insert(white, 2 * FEATHER_WIDTH, 2 * FEATHER_WIDTH)

    return alpha

# add a soft alpha and pad each image out to the final size
print("adding feather ...")
images = [image.bandjoin(soft_alpha(image.width, image.height))
          for image in images]

# composite all the images together at random positions
base = pyvips.Image.black(10000, 10000)
xes = [random.randint(0, base.width) for image in images]
yes = [random.randint(0, base.height) for image in images]
final = base.composite(images, 'over', x=xes, y=yes)

print("starting final write ...")
final.write_to_file(sys.argv[1])
```

That does a feathered join, which you might not need.

**Reply by zhusihan-python** [reply]
here is an example dataset with 11 rows * 9 cols = 99 images and a configure file which contains the tile locations: download [link](https://drive.google.com/drive/folders/1tXbDOmKCw8uGWGJl9SSzKtvu5ic625_T?usp=sharing)

**Reply by zhusihan-python** [reply]
as i tested: 99 2048 * 1536 jpg image composite to a jpg file cost 5.31s which is reasonable 
yet 1750 3072 * 2048 jpg file composite to a png(exceed the jpg size limitation) file cost 2332s which is insane

**Reply by jcupitt** [reply]
Yes, I designed composite thinking people might want to compose 3 or 4 images. The data structure (a simple linear list) does not scale well! It needs something better for large image sets.

**Reply by zhusihan-python** [reply]
what if i trans the final pyvips Image to numpy array, then try to find another way to write it. when i try to trans it by final.numpy() or np.asarray(final), i just got an out of memory error, is this for the final image is too large?
```
---------------------------------------------------------------------------
Error                                     Traceback (most recent call last)
Cell In[4], line 46
     44 print(f"starting final write composite cost {end1-start} seconds")
     45 # final.write_to_file(r"E:\projects\NeatSvtScanOptimize\data\20230329_1\captures\vips_mosaic.png")
---> 46 array = np.asarray(final)
     47 end2 = time.perf_counter()
     48 print("array shape", array.shape)

File D:\softwares\py3109\lib\site-packages\pyvips\vimage.py:1234, in Image.__array__(self, dtype)
   1208 """Conversion to a NumPy array.
   1209 
   1210 Args:
   (...)
   1229 See Also `Image.new_from_array` for the inverse operation. #TODO
   1230 """
   1231 import numpy as np
   1233 arr = (
-> 1234     np.frombuffer(self.write_to_memory(),
   1235                   dtype=FORMAT_TO_TYPESTR[self.format])
   1236     .reshape(self.height, self.width, self.bands)
   1237 )
   1239 if self.bands == 1:
   1240     # flatten single-band images
   1241     arr = arr.squeeze(axis=-1)

File D:\softwares\py3109\lib\site-packages\pyvips\vimage.py:944, in Image.write_to_memory(self)
    942 pointer = vips_lib.vips_image_write_to_memory(self.pointer, psize)
    943 if pointer == ffi.NULL:
--> 944     raise Error('unable to write to memory')
    945 pointer = ffi.gc(pointer, glib_lib.g_free)
    947 return ffi.buffer(pointer, psize[0])

Error: unable to write to memory
  vips_tracked: out of memory --- size == 0MB
vips_tracked: out of memory --- size == 0MB
vips_tracked: out of memory --- size == 0MB
...

```

​

**Reply by jcupitt** [reply]
Yes, the problem is assembly, not the final write.

**Reply by zhusihan-python** [reply]
i found that merge is much faster than compose, for the first 291 images, it works fine. but when it comes to the 292 image, got error too much overlap
```
Traceback (most recent call last):
  File "C:\Users\allen\Desktop\vips_merge.py", line 234, in <module>
    base = base.merge(images[i], "horizontal", -xes[i], -yes[i], mblend = 400)
  File "D:\softwares\py310\lib\site-packages\pyvips\vimage.py", line 1347, in call_function
    return pyvips.Operation.call(name, self, *args, **kwargs)
  File "D:\softwares\py310\lib\site-packages\pyvips\voperation.py", line 305, in call
    raise Error('unable to call {0}'.format(operation_name))
pyvips.error.Error: unable to call merge
  lrmerge: too much overlap
```
my code related to merge is like this
```
images = glob.glob(r"D:\图像测试数据\816\*.jpg")

print("loading images ...")
images = [pyvips.Image.new_from_file(filename, access="sequential") for filename in images]

base = images[0]
xes = []
yes = []

for k, v in total_config.items():
    xes.append(round(v[0]))
    yes.append(round(v[1]))

start0 = time.perf_counter()

for i in range(1, len(images)):
    if i >= 292:
        break
    if i <= y_squares-1:
        print(f"xes[{i}]", xes[i], yes[i])
        base = base.merge(images[i], "vertical", -xes[i], -yes[i], mblend = 400)
    elif (i//y_squares) % 2 != 0:
        index = 24*(2*(i//y_squares)+1)-1-i
        print("current i", i)
        print("index", index)
        print(f"xes[{index}]", xes[index], yes[index])
        base = base.merge(images[index], "horizontal", -xes[index], -yes[index], mblend = 400)
    else:
        print("current i", i)
        print(f"xes[{i}]", xes[i], yes[i])
        base = base.merge(images[i], "horizontal", -xes[i], -yes[i], mblend = 400)
```
some of the offsets output
```
index 264
xes[264] 30061 87243
current i 288
xes[288] 31906 14
current i 289
xes[289] 31947 3819
current i 290
xes[290] 31988 7599
current i 291
xes[291] 32027 11404
current i 292
xes[292] 32026 15166
```
does the x_offset has a limitation like over 32000?
if i ignore the rest images, i can get a result image like this, the red rect should be the 292 tile
![image](https://user-images.githubusercontent.com/23504604/232200746-9852eca7-f504-4f9b-bb24-5250f826d065.png)
may i just disable the overlap check in pyvips, do you got any suggestions sir

**Reply by jcupitt** [reply]
I would assemble the vertical strips separately with merge vertical, then join the set of strips with a horizontal merge. This will reduce the pipeline length and spread errors more evenly.

If you want to get fancy, assemble a set of 2x2 blocks to reduce your tiles by a factor of 2 in each dimension, then assemble those 2x2 blocks again into 4x4 blocks. This will have the best error distribution, but is a little harder to program.

**Reply by zhusihan-python** [reply]
can i just comment this few [lines](https://github.com/libvips/libvips/blob/c0a0630c6bc10420766a6f83adcdd02330fa859d/libvips/mosaicing/lrmerge.c#L849-L854) to get rid of  the overlap check.

i comment them and rebuild libvips 8.14.2, replace the old libvips refresh the environment envs, it seems not working, still got the same error
